proto/src/rr/dnssec/rdata/nsec3.rs

```diff
@@ -240,23 +240,34 @@ impl NSEC3 {
 }
 
 /// Read the RData from the given Decoder
-pub fn read(decoder: &mut BinDecoder, rdata_length: u16) -> ProtoResult<NSEC3> {
+pub fn read(decoder: &mut BinDecoder, rdata_length: Restrict<u16>) -> ProtoResult<NSEC3> {
     let start_idx = decoder.index();
 
-    let hash_algorithm = Nsec3HashAlgorithm::from_u8(decoder.read_u8()?)?;
-    let flags: u8 = decoder.read_u8()?;
+    let hash_algorithm = Nsec3HashAlgorithm::from_u8(decoder.read_u8()?.unverified())?;
+    let flags: u8 = decoder
+        .read_u8()?
+        .verify_unwrap(|flags| flags & 0b1111_1110 == 0)
+        .map_err(|flags| ProtoError::from(ProtoErrorKind::UnrecognizedNsec3Flags(flags)))?;
 
-    if flags & 0b1111_1110 != 0 {
-        return Err(ProtoErrorKind::UnrecognizedNsec3Flags(flags).into());
-    }
     let opt_out: bool = flags & 0b0000_0001 == 0b0000_0001;
-    let iterations: u16 = decoder.read_u16()?;
-    let salt_len: u8 = decoder.read_u8()?;
-    let salt: Vec<u8> = decoder.read_vec(salt_len as usize)?;
-    let hash_len: u8 = decoder.read_u8()?;
-    let next_hashed_owner_name: Vec<u8> = decoder.read_vec(hash_len as usize)?;
-
-    let bit_map_len = rdata_length as usize - (decoder.index() - start_idx);
+    let iterations: u16 = decoder.read_u16()?.unverified();
+    let salt_len = decoder
+        .read_u8()?
+        .map(|u| u as usize)
+        .verify_unwrap(|salt_len| *salt_len <= decoder.len())
```

Comment:
a potential excessive read

Reply:
Should there be a check for `rdata_length <= decoder.len()` and then compare `salt_len < rdata_length`? Because it should be error for `salt_len` and later `hash_len` to be over `rdata_length` long. Also `salt_len + hash_len < rdata_length`.

Reply:
Hm, this is a good call. `rdata_length <= decoder.len()` is checked in the higher up in resource.rs:303

These checks should be less than rdata_length though... and your recommendation is better on hash as well.

Reply:
Ok, rather than that calculation, I'm just using rdata_len.checked_sub(decoder.index() - start_idx) amount read from rdata...